ps3netsrv buffer overflow

[core.12076.zip](https://github.com/aldostools/webMAN-MOD/files/1713869/core.12076.zip)

Linux  4.15.1-2-ARCH
Installed from https://aur.archlinux.org/packages/ps3netsrv

ps3netsrv /home/user/PS3/ps3netsrv/                                                                                          
ps3netsrv build 20170310 (mod by aldostools)                                                                                                
Path: /home/user/PS3/ps3netsrv

Waiting for client...
Connection from 192.168.0.195                                                                                                               
open /GAMES/Heavy.Rain.USA/PS3_GAME/ICON0.PNG                                                                                               
Connection from 192.168.0.195                                                                                                               
building virtual iso...                                                                                                                     
*** buffer overflow detected ***: ps3netsrv terminated                                                                                      
Aborted (core dumped)

coredumpctl info 12076:
```
       Message: Process 12076 (ps3netsrv) of user 1000 dumped core.

                Stack trace of thread 12414:
                #0  0x00007f75b7a1f860 raise (libc.so.6)
                #1  0x00007f75b7a20ec9 abort (libc.so.6)
                #2  0x00007f75b7a62437 __libc_message (libc.so.6)
                #3  0x00007f75b7af239f __GI___fortify_fail_abort (libc.so.6)
                #4  0x00007f75b7af23d1 __fortify_fail (libc.so.6)
                #5  0x00007f75b7af0260 __chk_fail (libc.so.6)
                #6  0x00007f75b7aef426 __stpcpy_chk (libc.so.6)
                #7  0x0000565070fa65cf n/a (ps3netsrv)
                #8  0x0000565070fa717a n/a (ps3netsrv)
                #9  0x0000565070fa7a07 n/a (ps3netsrv)
                #10 0x0000565070fa7bb5 n/a (ps3netsrv)
                #11 0x0000565070fa44fc n/a (ps3netsrv)
                #12 0x00007f75b7fc008c start_thread (libpthread.so.0)
                #13 0x00007f75b7ae0e7f __clone (libc.so.6)

                Stack trace of thread 12076:
                #0  0x00007f75b7fc9cab accept (libpthread.so.0)
                #1  0x0000565070fa2df9 n/a (ps3netsrv)
                #2  0x00007f75b7a0bf4a __libc_start_main (libc.so.6)
                #3  0x0000565070fa32ca n/a (ps3netsrv)
```

coredumpctl gdb 12076 
and bt:
```
Reading symbols from /usr/bin/ps3netsrv...(no debugging symbols found)...done.
[New LWP 12414]
[New LWP 12076]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
Core was generated by `ps3netsrv /home/user/PS3/ps3netsrv/'.
Program terminated with signal SIGABRT, Aborted.
#0  0x00007f75b7a1f860 in raise () from /usr/lib/libc.so.6
[Current thread is 1 (Thread 0x7f75b639b700 (LWP 12414))]
(gdb) bt
#0  0x00007f75b7a1f860 in raise () from /usr/lib/libc.so.6
#1  0x00007f75b7a20ec9 in abort () from /usr/lib/libc.so.6
#2  0x00007f75b7a62437 in __libc_message () from /usr/lib/libc.so.6
#3  0x00007f75b7af239f in __fortify_fail_abort () from /usr/lib/libc.so.6
#4  0x00007f75b7af23d1 in __fortify_fail () from /usr/lib/libc.so.6
#5  0x00007f75b7af0260 in __chk_fail () from /usr/lib/libc.so.6
#6  0x00007f75b7aef426 in __stpcpy_chk () from /usr/lib/libc.so.6
#7  0x0000565070fa65cf in ?? ()
#8  0x0000565070fa717a in ?? ()
#9  0x0000565070fa7a07 in ?? ()
#10 0x0000565070fa7bb5 in ?? ()
#11 0x0000565070fa44fc in ?? ()
```

I just ran into this issue as well and came up with a quick workaround, deeper analysis and fix. Please reopen this issue until then.

When building the AUR package using makepkg I can reproduce the issue but not when building ps3netsrv manually. The difference is the build environment. I was able to work around this problem by disabling the buildflags and makeflags option.

Filling out the gaps in the traces by reproducing with a debug build:

coredump stack trace:
```
                #0  0x00007f0fbe393f25 raise (libc.so.6)
                #1  0x00007f0fbe37d897 abort (libc.so.6)
                #2  0x00007f0fbe3d7258 __libc_message (libc.so.6)
                #3  0x00007f0fbe467155 __GI___fortify_fail_abort (libc.so.6)
                #4  0x00007f0fbe467185 __fortify_fail (libc.so.6)
                #5  0x00007f0fbe4659d4 __chk_fail (libc.so.6)
                #6  0x00007f0fbe465238 __strcat_chk (libc.so.6)
                #7  0x000055c4fca40686 strcat (ps3netsrv)
                #8  0x000055c4fca41496 _ZN8VIsoFile5buildEPc (ps3netsrv)
                #9  0x000055c4fca41d39 _ZN8VIsoFile8generateEPcPKcS2_ (ps3netsrv)
                #10 0x000055c4fca41ee3 _ZN8VIsoFile4openEPKci (ps3netsrv)
                #11 0x000055c4fca3eaae process_open_cmd (ps3netsrv)
                #12 0x00007f0fbe5284cf start_thread (libpthread.so.0)
                #13 0x00007f0fbe4572d3 __clone (libc.so.6)
```

gdb backtrace:
```
#0  0x00007f0fbe393f25 in raise () from /usr/lib/libc.so.6
#1  0x00007f0fbe37d897 in abort () from /usr/lib/libc.so.6
#2  0x00007f0fbe3d7258 in __libc_message () from /usr/lib/libc.so.6
#3  0x00007f0fbe467155 in __fortify_fail_abort () from /usr/lib/libc.so.6
#4  0x00007f0fbe467185 in __fortify_fail () from /usr/lib/libc.so.6
#5  0x00007f0fbe4659d4 in __chk_fail () from /usr/lib/libc.so.6
#6  0x00007f0fbe465238 in __strcat_chk () from /usr/lib/libc.so.6
#7  0x000055c4fca40686 in strcat (__src=0x55c4fca426bd ";1", __dest=0x7f0fb8010211 "PS3_DISC.SFB") at /usr/include/bits/string_fortified.h:128
#8  VIsoFile::buildContent (this=this@entry=0x7f0fb8000c30, dirList=dirList@entry=0x7f0fb8000b60, joliet=joliet@entry=false) at ../src/VIsoFile.cpp:799
#9  0x000055c4fca41496 in VIsoFile::build (this=this@entry=0x7f0fb8000c30, inDir=inDir@entry=0x7f0fb8000bb0 <hidden>) at ../src/VIsoFile.cpp:1154
#10 0x000055c4fca41d39 in VIsoFile::generate (this=0x7f0fb8000c30, inDir=0x7f0fb8000bb0 <hidden>, volumeName=0x7f0fbdf0adb0 "PS3VOLUME", gameCode=0x7f0fbdf0ada0 <hidden>) at ../src/VIsoFile.cpp:1365
#11 0x000055c4fca41ee3 in VIsoFile::open (this=0x7f0fb8000c30, path=0x7f0fb8000bb0 <hidden>, flags=<optimized out>) at ../src/VIsoFile.cpp:1442
#12 0x000055c4fca3eaae in process_open_cmd (cmd=0x7f0fbdf0aea0, client=0x55c4fca45260 <clients>) at ../src/main.cpp:451
#13 client_thread (arg=0x55c4fca45260 <clients>) at ../src/main.cpp:1512
#14 0x00007f0fbe5284cf in start_thread () from /usr/lib/libpthread.so.0
#15 0x00007f0fbe4572d3 in clone () from /usr/lib/libc.so.6
```

Having nailed down the problem to:
https://github.com/aldostools/webMAN-MOD/blob/737a4c92044388e03d3ccb590c7e91f85323848a/_Projects_/ps3netsrv/src/VIsoFile.cpp#L799

This line even generates a warning:
```
In file included from /usr/include/string.h:494,
                 from ../src/VIsoFile.cpp:3:
In function ‘char* strcat(char*, const char*)’,
    inlined from ‘bool VIsoFile::buildContent(DirList*, bool)’ at ../src/VIsoFile.cpp:799:11:
/usr/include/bits/string_fortified.h:128:33: warning: ‘char* __builtin___strcat_chk(char*, const char*, long unsigned int)’ writing 3 bytes into a region of size 1 overflows the destination [-Wstringop-overflow=]
  128 |   return __builtin___strcat_chk (__dest, __src, __bos (__dest));
      |          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

This warning becomes a problem under certain build flags and thus has not necessarily led to problems before.

I'll shortly follow up with a pull request with a hotfix. While the problem lies deeper in the memory layout of the structs, I found it to be too much work to properly rework that. My fix will avoid the problem by doing this concatenation in a correctly allocated buffer as done in other places. Perhaps I'll rewrite server from scratch one day.

--- a/_Projects_/ps3netsrv/src/VIsoFile.cpp
+++ b/_Projects_/ps3netsrv/src/VIsoFile.cpp
@@ -793,19 +793,18 @@ bool VIsoFile::buildContent(DirList *dirList, bool joliet)
 
 			char *fileName = strrchr(fileList->path, '/') + 1;
 
+			char *s = new char[strlen(fileName) + 3];
+			strcpy(s, fileName);
+			strcat(s, ";1");
 			if (!joliet)
 			{
-				record->len_fi = strncpy_upper(&record->fi, fileName, MAX_ISONAME - 2) + 2;
-				strcat(&record->fi, ";1");
+				record->len_fi = strncpy_upper(&record->fi, s, MAX_ISONAME - 2);
 			}
 			else
 			{
-				char *s = new char[strlen(fileName) + 3];
-				strcpy(s, fileName);
-				strcat(s, ";1");
 				record->len_fi = utf8_to_ucs2((const unsigned char *)s, (uint16_t *)&record->fi, MAX_ISONAME/2) * 2;
-				delete[] s;
 			}
+			delete[] s;
 
 			record->len_dr = 0x27 + record->len_fi;
 			if (record->len_dr & 1)

